Mathlib/Algebra/Order/Hom/Basic.lean

```diff
@@ -118,3 +139,140 @@ theorem le_map_div_add_map_div [Group α] [AddCommSemigroup β] [LE β] [MulLEAd
+
+You should extend this class when you extend `GroupSeminorm`. -/
+@[to_additive]
+class GroupSeminormClass (F : Type _) (α β : outParam $ Type _) [Group α] [OrderedAddCommMonoid β]
+  extends MulLEAddHomClass F α β :=
+(map_one_eq_zero (f : F) : f 1 = 0)
+(map_inv_eq_map (f : F) (a : α) : f a⁻¹ = f a)
+
+/-- `AddGroupNormClass F α` states that `F` is a type of `β`-valued norms on the additive group
+`α`.
+
+You should extend this class when you extend `AddGroupNorm`. -/
+class AddGroupNormClass (F : Type _) (α β : outParam $ Type _) [AddGroup α] [OrderedAddCommMonoid β]
+  extends AddGroupSeminormClass F α β :=
+(eq_zero_of_map_eq_zero (f : F) {a : α} : f a = 0 → a = 0)
+
+/-- `GroupNormClass F α` states that `F` is a type of `β`-valued norms on the group `α`.
+
+You should extend this class when you extend `GroupNorm`. -/
+@[to_additive]
+class GroupNormClass (F : Type _) (α β : outParam $ Type _) [Group α] [OrderedAddCommMonoid β]
+  extends GroupSeminormClass F α β :=
+(eq_one_of_map_eq_zero (f : F) {a : α} : f a = 0 → a = 1)
+
+export AddGroupSeminormClass    (map_neg_eq_map)
+export GroupSeminormClass        (map_one_eq_zero map_inv_eq_map)
+export AddGroupNormClass        (eq_zero_of_map_eq_zero)
+export GroupNormClass            (eq_one_of_map_eq_zero)
+
+attribute [simp] map_one_eq_zero
+attribute [simp] map_neg_eq_map
+attribute [simp] map_inv_eq_map
+
+-- See note [lower instance priority]
+instance (priority := 100) AddGroupSeminormClass.to_ZeroHomClass [AddGroup α]
+  [OrderedAddCommMonoid β] [AddGroupSeminormClass F α β] : ZeroHomClass F α β :=
+{ ‹AddGroupSeminormClass F α β› with }
+
+section GroupSeminormClass
+variable [Group α] [OrderedAddCommMonoid β] [GroupSeminormClass F α β] (f : F) (x y : α)
+
+@[to_additive] lemma map_div_le_add : f (x / y) ≤ f x + f y :=
+by rw [div_eq_mul_inv, ←map_inv_eq_map f y]; exact map_mul_le_add _ _ _
+
+@[to_additive] lemma map_div_rev : f (x / y) = f (y / x) := by rw [←inv_div, map_inv_eq_map]
+
+@[to_additive] lemma le_map_add_map_div' : f x ≤ f y + f (y / x) :=
+by simpa only [add_comm, map_div_rev, div_mul_cancel'] using map_mul_le_add f (x / y) y
+
+end GroupSeminormClass
+
+example [OrderedAddCommGroup β] : OrderedAddCommMonoid β := inferInstance
+
+@[to_additive] lemma abs_sub_map_le_div [Group α] [LinearOrderedAddCommGroup β]
+  [GroupSeminormClass F α β] (f : F) (x y : α) : |f x - f y| ≤ f (x / y) := by
+  rw [abs_sub_le_iff, sub_le_iff_le_add', sub_le_iff_le_add']
+  exact ⟨le_map_add_map_div _ _ _, le_map_add_map_div' _ _ _⟩
+
+@[to_additive] -- See note [lower instance priority]
+instance (priority := 100) GroupSeminormClass.to_NonNegHomClass [Group α]
+  [LinearOrderedAddCommMonoid β] [GroupSeminormClass F α β] :
+  NonNegHomClass F α β :=
+{ ‹GroupSeminormClass F α β› with
+  map_nonneg := fun f a ↦ (nsmul_nonneg_iff two_ne_zero).1 $
+    by rw [two_nsmul, ←map_one_eq_zero f, ←div_self' a]; exact map_div_le_add _ _ _ }
```

Comment:
There seems to be a bug with `to_additive` here.
```
application type mismatch
  HVAdd.hVAdd
argument has type
  HSMul ℕ ((fun x => β) a) ((fun x => β) a)
but function has type
  [self : HVAdd ℕ ((fun x => β) a) ((fun x => β) a)] → ℕ → (fun x => β) a → (fun x => β) a
```

Reply:
Can you merge master and check if the error is still there?

Reply:
#1332 should fix this error. There are still type-class errors in the rest of the file.

Reply:
#1332 is now merged